### ISSUE TITLE
(Draft) Trade

### ISSUE THREAD

**Issue by julkascript** [julkascript/cardflow#116]
General requirements:
- A user (Requester) can request trade to another user (Responder)
- The Requester can select what they require from a predefined list, which is generated from the Responder's listings. In other words, Requesters can only request what Responders have listed in their "inventory".
- The Responder can Accept, Decline, or Negotiate
  - The Negotiation is like a "counter-offer". The Responder and the requester switch places. This can be done multiple times.
-  Users can also additional options, which for now can only be "cash". For example, a user can trade "card X" for "card Y" and 20 BGN.

**Comment by RyotaMitaraiWeb** [comment]
I am seeking a clarification on some behavior, because the way it's implemented on the backend and the way it's portrayed in the wireframes is confusing me:

- reading the requirements and the wireframes (in particular the chat ones), I assume that one should be able to change the other user's listings. In addition, when initiating a trade offer, you would obviously need to do that as well so that the other user knows what you want.
- However, on the backend, sending a ``PATCH`` request to the trade endpoint gives me a 400 error when I try to modify the other's listings, stating that I am not allowed to change the other user's listings
- On top of this, the update *actually* succeeds despite the 400 error (this one feels like a bug)

Am I understanding / doing something wrong? If so, what would be the correct behavior / course of action in that case?

**Comment by julkascript** [comment]
> I am seeking a clarification on some behavior, because the way it's implemented on the backend and the way it's portrayed in the wireframes is confusing me:
> 
> * reading the requirements and the wireframes (in particular the chat ones), I assume that one should be able to change the other user's listings. In addition, when initiating a trade offer, you would obviously need to do that as well so that the other user knows what you want.
> * However, on the backend, sending a `PATCH` request to the trade endpoint gives me a 400 error when I try to modify the other's listings, stating that I am not allowed to change the other user's listings
> * On top of this, the update _actually_ succeeds despite the 400 error (this one feels like a bug)
> 
> Am I understanding / doing something wrong? If so, what would be the correct behavior / course of action in that case?

Yes, you are correct. A PATCH endpoint should allow changes of the listings on both sides. However, the listings should still be owned by the other user, you should not be able to select listings of other users. Maybe it’s a bug. @dinocom33 can you double-check?

**Comment by dinocom33** [comment]
If you can fix it, do it or raise an issue. Which non registered actions do you mean?

**Comment by RyotaMitaraiWeb** [comment]
> If you can fix it, do it or raise an issue. Which non registered actions do you mean?

there seem to be some actions called "accept", "reject", "negotiate", and "counter_offer" They are not generated in the Swagger client and making requests to those gives me 404 errors. Only when I explicitly register them (e.g. ``path('<int:pk>/negotiate/', TradeListingViewSet.as_view({'post': 'negotiate'})),``) do the actions actually do what they are supposed to.

(based on reading the DRF docs, actions are supposed to be routable, that's why I initially asked my question)

**Comment by dinocom33** [comment]
> > If you can fix it, do it or raise an issue. Which non registered actions do you mean?
> 
> there seem to be some actions called "accept", "reject", "negotiate", and "counter_offer" They are not generated in the Swagger client and making requests to those gives me 404 errors. Only when I explicitly register them (e.g. `path('<int:pk>/negotiate/', TradeListingViewSet.as_view({'post': 'negotiate'})),`) do the actions actually do what they are supposed to.
> 
> (based on reading the DRF docs, actions are supposed to be routable, that's why I initially asked my question)

All endpoints are refactored and regarding issue  ```Trade Statuses are not accurate #175```
